src/api_client.rs

```diff
@@ -111,12 +112,40 @@ impl Client {
         self.get("/eth/v1/beacon/genesis").await
     }
 
-    pub async fn get_state_root(id: StateId) -> Result<Root, Error> {
-        unimplemented!("")
+    pub async fn get_state_root(&self, state_id: StateId) -> Result<Root, Error> {
+        let mut stub = "";
+
+        match state_id {
+            StateId::Finalized => stub = "finalized",
+            StateId::Head => stub = "head",
+            StateId::Genesis => stub = "genesis",
+            StateId::Root(_) => stub = "not implemented",
+            StateId::Slot(_) => stub = "not implemented",
+            StateId::Justified => stub = "justified",
+        };
+
+        let path = format!("eth/v1/beacon/states/{}/root", &stub);
```

Comment:
so a different way to do this that avoids having to copy this code each time we use it is to move the logic to the type, in this case `StateId`

we can then identify the common behavior here (aka "rendering this as a string") and implement a trait for that behavior

rust already has many built-in traits for all sorts of conversions, like ToString: https://doc.rust-lang.org/std/string/trait.ToString.html and if you read the docs then it requests you to implement `fmt::Display`

that gives an impl of `to_string` for free and the way you have set up this `format!` call it should call the right logic for you

you don't even need to pass `state_id` by reference, you can just drop it into the call to `format!`.

so I would implement `std::fmt::Display` for `StateId` and return the correct string data for each variant

then you can delete this block here and the one below

Reply:
ok, yea that seems obvious now. Fixed!